core/templates/cowdata.h

```diff
@@ -293,143 +304,137 @@ void CowData<T>::_unref() {
 }
 
 template <typename T>
-typename CowData<T>::USize CowData<T>::_copy_on_write() {
-	if (!_ptr) {
-		return 0;
+Error CowData<T>::_fork_allocate(USize p_size) {
+	if (p_size == 0) {
+		// Wants to clean up.
+		_unref();
+		return OK;
 	}
 
-	SafeNumeric<USize> *refc = _get_refcount();
-
-	USize rc = refc->get();
-	if (unlikely(rc > 1)) {
-		/* in use by more than me */
-		USize current_size = *_get_size();
+	USize alloc_size;
+	ERR_FAIL_COND_V(!_get_alloc_size_checked(p_size, &alloc_size), ERR_OUT_OF_MEMORY);
 
-		uint8_t *mem_new = (uint8_t *)Memory::alloc_static(_get_alloc_size(current_size) + DATA_OFFSET, false);
-		ERR_FAIL_NULL_V(mem_new, 0);
+	const USize prev_size = size();
 
-		SafeNumeric<USize> *_refc_ptr = _get_refcount_ptr(mem_new);
-		USize *_size_ptr = _get_size_ptr(mem_new);
-		T *_data_ptr = _get_data_ptr(mem_new);
+	if (!_ptr) {
+		// We had no data before; just allocate a new array.
+		const Error error = _alloc(alloc_size);
+		if (error) {
+			return error;
+		}
+	} else if (_get_refcount()->get() == 1) {
+		// Resize in-place.
+		// Mind that this case is not just an optimization, but required,
+		//  as some callers depend on _copy_on_write() calls not changing
+		//  the pointer after the first fork (e.g. mutable iterators).
+		if (p_size == prev_size) {
+			// We can shortcut here; we don't need to do anything.
+			return OK;
+		}
 
-		new (_refc_ptr) SafeNumeric<USize>(1); //refcount
-		*(_size_ptr) = current_size; //size
+		// Destroy extraneous elements.
+		if constexpr (!std::is_trivially_destructible_v<T>) {
+			for (USize i = prev_size; i > p_size; i--) {
+				_ptr[i - 1].~T();
+			}
+		}
 
-		// initialize new elements
-		if constexpr (std::is_trivially_copyable_v<T>) {
-			memcpy((uint8_t *)_data_ptr, _ptr, current_size * sizeof(T));
-		} else {
-			for (USize i = 0; i < current_size; i++) {
-				memnew_placement(&_data_ptr[i], T(_ptr[i]));
+		if (alloc_size != _get_alloc_size(prev_size)) {
+			const Error error = _realloc(alloc_size);
+			if (error) {
+				// Out of memory; the current array is still valid though.
+				return error;
 			}
 		}
+	} else {
+		// Resize by forking.
+
+		// Create a temporary CowData to hold ownership over our _ptr.
+		// It will be used to copy elements from the old buffer over to our new buffer.
+		// At the end of the block, it will be automatically destructed by going out of scope.
+		const CowData prev_data;
+		prev_data._ptr = _ptr;
+		_ptr = nullptr;
 
-		_unref();
-		_ptr = _data_ptr;
+		const Error error = _alloc(alloc_size);
+		if (error) {
+			return error;
+		}
 
-		rc = 1;
+		// Copy over elements.
+		const USize copied_element_count = MIN(prev_size, p_size);
+		if (copied_element_count > 0) {
+			if constexpr (std::is_trivially_copyable_v<T>) {
+				memcpy((uint8_t *)_ptr, (uint8_t *)prev_data._ptr, copied_element_count * sizeof(T));
+			} else {
+				for (USize i = 0; i < copied_element_count; i++) {
+					memnew_placement(&_ptr[i], T(prev_data._ptr[i]));
```

Comment:
As we will destroy prev_data._ptr[i] after this, could we not move them in ? prev_data being const, the access will be `const T&` (I guess), so it will trigger the copy constructor. There is probably some cases where the move constructor could be faster.

Reply:
`prev_data._ptr[i]` will likely not be destroyed. It's copying over elements because we (the CowData doing the copying) is not the sole owner of the backing array. The case where we are the sole owner, and may destroy the old array, is the realloc case above :)

Reply:
Oh yes it's copying, the old one is still valid, my bad. You can close this.